Remove wp-data & wc-blocks-checkout dependencies from OA script

tomalec opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

- Remove wp-data & wc-blocks-checkout dependencies from OA script
   Resolve conditional dependency in JS.

Closes: https://github.com/woocommerce/woocommerce/issues/43829


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. (developer only) Check that order-attribution js does not have `'wp-data', 'wc-blocks-checkout'` as dependencies.
1. Check live site:
   1. Ensure the mini-cart block is removed from your header.
      ![image](https://github.com/woocommerce/woocommerce/assets/17435/606978de-6f2b-4e3d-a5e2-bf29283800de)

   2. Load the page and ensure the `wc-blocks-checkout` script is not loaded.
      Call from the browser dev tools `document.querySelector('script#wc-blocks-checkout-js')` - should return `null`
2. Check testing instructions for the following for the regression:
   1. https://github.com/woocommerce/woocommerce/pull/39701
   3. https://github.com/woocommerce/woocommerce/pull/43012
<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

tomalec committed:
Remove `wp-data` & `wc-blocks-checkout` dependencies from OA script

Resolve conditional dependency in JS.

Fixes https://github.com/woocommerce/woocommerce/issues/43829

Co-authored-by: Seghir Nadir <nadir.seghir@gmail.com>

tomalec committed:
Unsubscribe wp.data listener in OA

to prevent infinite loop.

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

tomalec committed:
Subscribe only to `'wc/store/checkout'` changes

tomalec committed:
Add the changelog entry

tomalec committed:
Merge branch 'trunk' into fix/43829-oa-blocks

opr commented on plugins/woocommerce/client/legacy/js/frontend/order-attribution.js:
```diff
@@ -111,4 +114,28 @@
+		) {
+			const unsubscribe = wp.data.subscribe( function () {
+				const checkoutDataStore = wp.data.select( CHECKOUT_STORE_KEY );
+				if ( undefined !== checkoutDataStore && checkoutDataStore.getOrderId() !== 0 ) {
```
```suggestion
				if ( 'undefined' !== typeof checkoutDataStore && checkoutDataStore.getOrderId() !== 0 ) {
```

I prefer this way of doing it as it's a little safer.

tomalec replied:
`checkoutDataStore` constant is declared just the line above, so IMHO typeof is not needed.

Also, do we even need that check? We subscribed to `CHECKOUT_STORE_KEY` store changes, so `CHECKOUT_STORE_KEY` store should be guaranteed to exist in the callback, shouldn't it?

tomalec replied:
I did a quick micro-perf check, and it seems there is no significant difference between them. https://jsperf.app/nofavo

Using `typeof` here would make lines 127 and below "safe" in the sense they will run without visible error when line 126 is lost (for example, in a bad merge).
IMHO, this is actually **not desired**. If the line 126 is lost, the code should fail fast. So we could identify the bad merge or other bug before shipping it to production.

tomalec replied:
I removed this condition altogether in https://github.com/woocommerce/woocommerce/pull/43857/commits/f3abaf0304f75d5a0edf2a4e7709cd344ef4bfd3

We don't need to check if there is an order being processed in the checkout.
To add extension data, we only need the store to finally be there. That should be guaranteed by the  `subscribe` itself.

opr replied:
@tomalec I appreciate how deeply you considered this! Kudos. 🙏🏼 The result is better for sure.

opr reviewed (changes requested):
From the blocks point of view this is working nicely.

Just to recap: I removed the mini cart block from my site and tested before this PR. There were 55 JS scripts loaded, `wc-blocks-checkout` was one of them. The only dependent listed for this was `wc-order-attribution`. `wp-data` was also loaded.

After applying this PR, the total number of scripts loaded was 13, `wc-blocks-checkout` and `wp-data` were not included.

Do you think updating the testing instructions to include something like:

1. Ensure the mini-cart block is removed from your header.
2. Load the page and ensure the `wc-blocks-checkout` script is not loaded.

would be useful? Maybe mark it as "developer only" instructions?

tomalec commented:
Thank you for the fast review and testing.

I added the testing instructions as you suggested.

tomalec committed:
Remove redundant checks, update OA in checkout ASAP

Related to https://github.com/woocommerce/woocommerce/pull/43857#discussion_r1462134242

opr reviewed (approved):
Thanks for adding the test instructions, could we just specify checking the script dependencies is "dev only" so globalstep don't have to figure out how to do that?

tomalec committed:
Use `window.` for `wp` global.

tomalec commented:
Speaking of being super safe I've added [one more tweak](https://github.com/woocommerce/woocommerce/pull/43857/commits/5685035e694bc22378a4fccdefbf9d63b4c600b4)  to use `window.wp.` instead of `wp.` 
It was still after checking that `window.wp` is there, but it's always safer to use the same thing and be kind to linters and bundlers, as the file would use only one global.

tjcafferkey commented:
I've done some testing:

* The steps in this PR test well 👍🏻 
* https://github.com/woocommerce/woocommerce/pull/39701 tests well with no regressions for me 👍🏻 
* https://github.com/woocommerce/woocommerce/pull/43012 doesn't test well for me, perhaps I'm missing something but I've put some notes below 👎🏻 

**Fix Consent API integration with OSA (trunk-based)**

"Denied by default" step didn't return `false` values for:

>wc_order_attribution.params.allowTracking should be set to false and no order attribution data should be collected.

<img width="1472" alt="Screenshot 2024-01-24 at 07 19 33" src="https://github.com/woocommerce/woocommerce/assets/8639742/81be9654-675b-4d04-b719-345825aab9cd">

I also didn't have a `/checkout` request for the checkout block steps, for some reason. Perhaps I'm doing something obviously wrong but I've also asked about this in the Team Rubik channel. I stopped testing this PR until I've resolved these issues if its user-error on my end.

tomalec commented:
I replicated the `opt-in` + `wc_order_attribution.params.allowTracking == true` with Complianz set to UK mode, and with WP Consent API plugin deactivated.
![image](https://github.com/woocommerce/woocommerce/assets/17435/fb61609c-525e-4551-b267-6319b515951d)

tomalec commented:
If the problem is with `wc_order_attribution.params.allowTracking == true`, then it's most likely something unrelated to the changes from this PR, as none of the changed lines affect the flow that sets `allowTracking`.

I'm merging this PR to fix https://github.com/woocommerce/woocommerce/issues/43829 in the upcoming release. 
If the problem with `allowTracking` still happens with the WP Consent API installed, then it could be tracked as a separate issue.
